filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,211 @@
+			return
+		}
+	}()
+
+	input := &Input{
+		config:        config,
+		rawConfig:     cfg,
+		started:       false,
+		outlet:        out,
+		consumerGroup: consumerGroup,
+		kafkaContext:  kafkaContext,
+		kafkaCancel:   kafkaCancel,
+		log:           logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+func (p *Input) newConsumerGroup() (sarama.ConsumerGroup, error) {
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(p.config.Hosts, p.config.GroupID, nil)
+	return consumerGroup, err
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (p *Input) Run() {
+	if !p.started {
+		// Track errors
+		go func() {
+			for err := range p.consumerGroup.Errors() {
+				p.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: p}
+
+				err := p.consumerGroup.Consume(p.kafkaContext, p.config.Topics, handler)
+				if err != nil {
+					p.log.Errorw("Kafka consume error", "error", err)
+				}
+			}
+		}()
+		p.started = true
+	}
+}
+
+// Wait shuts down the Input by cancelling the internal context.
+func (p *Input) Wait() {
+	p.Stop()
+}
+
+// Stop shuts down the Input by cancelling the internal context.
+func (p *Input) Stop() {
+	p.kafkaCancel()
+}
+
+func arrayForKafkaHeaders(headers []*sarama.RecordHeader) []interface{} {
+	array := []interface{}{}
+	for _, header := range headers {
+		array = append(array, common.MapStr{
+			"key":   header.Key,
+			"value": header.Value,
+		})
+	}
+	return array
+}
+
+type groupHandler struct {
+	input *Input
+}
+
+func (h groupHandler) createEvent(
+	sess sarama.ConsumerGroupSession,
+	claim sarama.ConsumerGroupClaim,
+	message *sarama.ConsumerMessage,
+) *util.Data {
+	data := util.NewData()
+	data.Event = beat.Event{
+		Timestamp: time.Now(),
+	}
+	eventFields := common.MapStr{
+		"message": string(message.Value),
+	}
+	kafkaMetadata := common.MapStr{
+		"topic":     claim.Topic(),
+		"partition": claim.Partition(),
+		"offset":    message.Offset,
+		"key":       message.Key,
+	}
+	version, versionOk := h.input.config.Version.Get()
+	if versionOk && version.IsAtLeast(sarama.V0_10_0_0) {
+		data.Event.Timestamp = message.Timestamp
+		if !message.BlockTimestamp.IsZero() {
+			kafkaMetadata["block_timestamp"] = message.BlockTimestamp
+		}
+	}
+	if versionOk && version.IsAtLeast(sarama.V0_11_0_0) {
+		kafkaMetadata["headers"] = arrayForKafkaHeaders(message.Headers)
+	}
+	eventFields["kafka"] = kafkaMetadata
+	data.Event.Fields = eventFields
+	return data
+}
+
+func (groupHandler) Setup(session sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+func (groupHandler) Cleanup(_ sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+func (h groupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for msg := range claim.Messages() {
+		event := h.createEvent(sess, claim, msg)
+		h.input.outlet.OnEvent(event)
+		sess.MarkMessage(msg, "")
```

Comment:
we really need to pass `util.Data`? Need to check if we can ignore the outleter and util.Data in new inputs.

Marking a message here does not give us end to end ACK. libbeat has support for end to end ACK, but filebeat architecture somehow conceals this. This is somewhat ok for the file input, but no other inputs. The way it is written here we are subject to data loss if events get not ACKed before filebeat shuts down.